### ISSUE TITLE
Type inference incorrect with @HasQualifierParameter

### ISSUE THREAD

**Issue by jwaataja** [t-rasmud/checker-framework#154]
This also applies to [this branch](https://github.com/smillst/checker-framework/tree/implicitTypeParameter).

Writing `@HasQualifierParameter` on a class makes the default annotation for that class the polymorphic qualifier for that type system. This seems to override local type inference so that more annotations need to be written than should be necessary. Consider this example,
```java
import org.checkerframework.checker.determinism.qual.*;
import org.checkerframework.framework.qual.*;

@HasQualifierParameter(NonDet.class)
public class D {
    public static void f(@Det D a) {
        D b = a;
    }
}
```
which gives the error
```
D.java:7: error: [assignment.type.incompatible] incompatible types in assignment.
        D b = a;
              ^
  found   : @Det D
  required: @PolyDet D
```
It should be able to infer that `d` is of type `@Det D`.

There seem to be further issues when a class that has `@HasQualifierParameter` extends one that does not. Here's an example,
```java
import org.checkerframework.checker.determinism.qual.*;
import org.checkerframework.framework.qual.*;

public class C {
    public static void f(@Det D a) {
        C b = a;
        @Det C c = b;
    }
}

@HasQualifierParameter(NonDet.class)
class D extends C {
}
```
which gives this error
```
C.java:7: error: [assignment.type.incompatible] incompatible types in assignment.
        @Det C c = b;
                   ^
  found   : @NonDet C
  required: @Det C
```

Here, the type of `b` is neither `@PolyDet` nor `@Det` like it would be expected, but it is actually `@NonDet`.

**Comment by smillst** [comment]
The first error is as excepted.  The second example is a bug.  I'll take a look.

**Comment by smillst** [comment]
Fixed via 51d21188f3bbeb161e280193c5dd5015abd67830.  (I merged the implicitTypeParameter branch into nondet-checker branch.)
